chore: implement some basic tests.

peterhuene opened the pull request:
This PR implements some basic test coverage for the underlying default storage implementation and via the integration with `reqwest`.

peterhuene committed:
chore: implement some basic tests.

This commit implements some basic test coverage for the underlying
default storage implementation and via the integration with `reqwest`.

peterhuene commented:
Need to guard the storage tests against the `smol` feature.

peterhuene committed:
chore: fix test build.

peterhuene committed:
chore: update README.

acfoltzer commented on src/storage/default.rs:
```diff
@@ -479,3 +479,132 @@ impl DefaultCacheStorageInner {
+        ) -> Poll<Option<Result<Frame<Bytes>, Self::Error>>> {
+            self.project().body.poll_frame(cx).map_err(io::Error::other)
+        }
+    }
```
I recommend pulling in `http_body_util` rather than having to worry about our own `Body` implementation with all the complications of pinning and such: https://docs.rs/http-body-util/latest/http_body_util/struct.Full.html

peterhuene replied:
Thanks for letting me know about `http_body_util`. I ended up removing some unnecessary code as a result (in addition to this body impl).

acfoltzer commented on crates/reqwest/src/lib.rs:
```diff
@@ -211,3 +211,307 @@ impl<S: CacheStorage> reqwest_middleware::Middleware for Cache<S> {
+            "22b8d362b2e8064356915b1451f630d1d920b427d3b2f9b3432fbf4c03d94184";
+
+        let dir = tempdir().unwrap();
+        let revalidated = Arc::new(AtomicBool::new(false));
```
Since this is a one-off test and not a tight loop where performance is critical, I would use a boring but simpler `Mutex` rather than introducing the complication of atomics.

peterhuene replied:
I personally don't feel that using an atomic is complicated vs. a mutex with scoped locking. If you feel strongly about it, I can change it, otherwise I'd prefer the atomic.

acfoltzer replied:
I do think that unless it's necessary, seeing stuff like `Ordering::SeqCst` makes things unnecessarily intimidating to folks who haven't worked with atomics before.

acfoltzer commented on crates/reqwest/src/lib.rs:
```diff
@@ -211,3 +211,307 @@ impl<S: CacheStorage> reqwest_middleware::Middleware for Cache<S> {
+    #[tokio::test]
+    async fn max_age() {
+        const BODY: &str = "hello world!";
+        const DIGEST: &str = "3aa61c409fd7717c9d9c639202af2fae470c0ef669be7ba2caea5779cb534e9d";
```
Notes for how these `DIGEST` constants are derived for this and the other tests would be useful

acfoltzer commented on crates/reqwest/src/lib.rs:
```diff
@@ -211,3 +211,307 @@ impl<S: CacheStorage> reqwest_middleware::Middleware for Cache<S> {
+            response.headers().get(header::CACHE_CONTROL).unwrap(),
+            "no-store"
+        );
+        assert_eq!(response.headers().get(X_CACHE_LOOKUP).unwrap(), "MISS");
+        assert_eq!(response.headers().get(X_CACHE).unwrap(), "MISS");
+        assert!(response.headers().get(X_CACHE_DIGEST).is_none());
```
Definitely beside the point of this PR, but we should be using `cache-status` instead of `x-cache` headers unless needed for compatibility with another system; see https://www.rfc-editor.org/rfc/rfc9211.html

peterhuene replied:
Might be worth looking into making use of; the upstream crate this one was inspired from also uses the same extension headers, which aren't uncommon in the wild, but they definitely lack the standards track that `Cache-Status` has.

acfoltzer commented on crates/reqwest/src/lib.rs:
```diff
@@ -211,3 +211,307 @@ impl<S: CacheStorage> reqwest_middleware::Middleware for Cache<S> {
+        assert!(response.headers().get(X_CACHE_DIGEST).is_none());
+        assert_eq!(response.text().await.unwrap(), BODY);
+
+        // Ensure no content directory
```
What does this mean?

peterhuene replied:
Should read `ensure no content entry exists` (will update).

peterhuene committed:
chore: code review feedback.

* Use `http_body_util` where appropriate rather than manual `Stream`
  implementations.
* Remove no longer needed `HttpBody` trait.
* Address code review comments.

peterhuene committed:
chore: fix `smol` feature build.

peterhuene committed:
chore: remove use of `Full`.

With how `CacheBody` is now implemented, it can be backed by `String`,
so `Full` isn't necessary.

peterhuene committed:
chore: code review feedback.

Remove the use of atomics in the tests.

peterhuene committed:
chore: remove unused dep.

acfoltzer reviewed (approved):
Thanks!
